### DISCUSSION TITLE
FR: bidirectional links

### DISCUSSION THREAD

**Discussion by quesada** [zim-desktop-wiki/zim-desktop-wiki#1838] [Ideas]
Because most of us use zim to explore our thoughts, bidirectional links make more sense than monolinks (like the web ones)
More on this:
https://handbook.athensresearch.org/community/athens-guide/feature-list/bidirectional-links

Another good write up on the value of bidirectional links:
https://news.ycombinator.com/item?id=26316793

Zim is far more agile than any electron abomination, and it feels to me that adding bidirectional links would not be a lot of work (I might be wrong). This would take Zim to the forefront of 'research' in notetaking (note Athens is a Ycombinator company, which usually means it's finely tuned into the groupthink of movers and shakers in Sillicon Valley, FWIW).

**Comment by sojusnik** [comment]
Isn't this feature already covered with backlinks?

**Reply by quesada** [reply]
I used to see backlinks on a bar at the bottom (not seening it in 0.74? Where did it go?). But these bidirectional links are different. It's a foundational difference.It means that every thought is connected to others both ways. 
The reason I think this would be an easy feature to add is that it would simply have to 'add more' of what we have (links).

**Reply by sojusnik** [reply]
Backlinks are still there, just enable them in Plugins and specify the position to be at the bottom.

Taken from [Athens manual](https://handbook.athensresearch.org/community/athens-guide/feature-list/bidirectional-links):
> When you create a bidirectional link in Athens, you're creating two pages that are aware of each other. A links to B through what looks like a one-way link, but then B points back to A via a list of references.
…
As mentioned above, the links you create between pages are bidirectional, which means they appear on both "sides" of the link.
On the page being linked to, these links appear in the Linked References section below the outliner.

As I understand, the "Linked References" section is our backlinks, isn't it? So those pages are always connected "in both ways".

**Reply by introt** [reply]
The "Linked References" section seems to be "our backlinks on steroids", eg. listing not only the pages that link to this one, but showing the surrounding context as well. I'd say #1819 is at the very least related to this, but a full implementation would require many of the things listed here: https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/1818#discussioncomment-1815552

**Reply by sojusnik** [reply]
Ok, so it's the additional context you see. Sounds helpful, so I'm in!

**Reply by quesada** [reply]
@sojusnik I'm not an Athens user, but what I mean is that when you create a link in zim to a page, it would create another link on that destination page linking back. I think this is how it works in athens. That has consequences for the linkmap, any link analysis tool working on top of it, etc

This is why I suspect that adding the feature (with a checkbox on the floating window for 'create link' that would say 'bidirectional') should not be too complex.

**Reply by quesada** [reply]
@introt the listing is one thing (zim does it), but the creation of the link is a different topic. Zim doesn't create bidirectional links. And at least the way I use it to think, bidirectional links are more useful.

**Reply by introt** [reply]
> On the page being linked to, these links appear in the Linked References section below the outliner.

Nothing in the manual suggests they're doing anything besides the listing. I think the confusion stems from conflating links (text on a page/node) with links between pages (edges between nodes). "Bidirectional links" mean that for every edge stemming from a link in a source page, there's another one back from the destination page; the edges are created in pairs. This is different from creating a "concrete" text-on-a-page link in the destination, which (in addition to showing up as another arrow(head) on the linkmap) could be achieved with "monolinks".

Unlike Athens, which uses a graph database for _graphs_ (page nodes and the edges between them), Zim stores the concrete links in text files and both edges in the `index.db` database file. Paraphrasing from the writeup, "you can't really implement this with text files" - which is why such a database exists.
